src/backend/columnar/columnar_tableam.c

```diff
@@ -291,6 +292,22 @@ columnar_getnextslot(TableScanDesc sscan, ScanDirection direction, TupleTableSlo
  */
 static ItemPointerData
 row_number_to_tid(uint64 rowNumber)
+{
+	ErrorIfInvalidRowNumber(rowNumber);
+
+	ItemPointerData tid = { 0 };
+	ItemPointerSetBlockNumber(&tid, rowNumber / VALID_ITEMPOINTER_OFFSETS);
+	ItemPointerSetOffsetNumber(&tid, rowNumber % VALID_ITEMPOINTER_OFFSETS +
+							   FirstOffsetNumber);
+	return tid;
+}
+
+
+/*
+ * ErrorIfInvalidRowNumber errors out if given rowNumber is invalid.
+ */
+static void
+ErrorIfInvalidRowNumber(uint64 rowNumber)
```

Comment:
What about very large row numbers that can't be represented?

Reply:
we handle large row numbers in `ErrorIfInvalidRowNumber` too, in `else if (rowNumber > COLUMNAR_MAX_ROW_NUMBER)` block

Reply:
Oh, that part happened to be perfectly covered by the diff viewer.